### PR TITLE
adding KID metric

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Implements kernel inception distance (KID).
Should be merged after #299

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/image/kid.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        feature: Union[str, int, torch.nn.Module] = 2048,
+        subsets: int = 100,
+        subset_size: int = 1000,
+        degree: int = 3,
+        gamma: Optional[float] = None,
+        coef: float = 1.0,
+        compute_on_step: bool = False,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None
```
Comment (justusschock) [inline torchmetrics/image/kid.py:185]: nit: Can you give the type of the `Callable` inputs and outputs? e.g for str as input and int as output it would be `Callable[[str], int]`

File: torchmetrics/image/kid.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+        )
+
+        if isinstance(feature, (str, int)):
+            if not _TORCH_FIDELITY_AVAILABLE:
+                raise ValueError(
+                    'KID metric requires that Torch-fidelity is installed.'
+                    'Either install as `pip install torchmetrics[image-quality]`'
+                    ' or `pip install torch-fidelity`'
+                )
+            valid_int_input = ['logits_unbiased', 64, 192, 768, 2048]
```
Comment (justusschock) [inline torchmetrics/image/kid.py:206]:
```suggestion
            valid_int_input = ('logits_unbiased', 64, 192, 768, 2048)
```

use immutables whenever possible

File: torchmetrics/image/kid.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            if feature not in valid_int_input:
+                raise ValueError(
+                    f'Integer input to argument `feature` must be one of {valid_int_input},'
+                    f' but got {feature}.'
+                )
+
+            self.inception = NoTrainInceptionV3(name='inception-v3-compat', features_list=[str(feature)])
```
Comment (justusschock) [inline torchmetrics/image/kid.py:213]: Based on this: Does it make sense to also support a list of features?

File: torchmetrics/image/kid.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+        elif isinstance(feature, torch.nn.Module):
+            self.inception = feature
+        else:
+            raise ValueError('Got unknown input to argument `feature`')
```
Comment (justusschock) [inline torchmetrics/image/kid.py:217]:
```suggestion
            raise TypeError('Got unknown input to argument `feature`')
```

also reflect this in docstrings :)

File: torchmetrics/image/kid.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+        if not (isinstance(subset_size, int) and subset_size > 0):
+            raise ValueError("Argument `subset_size` expected to be integer larger than 0")
+        self.subset_size = subset_size
+
+        if not (isinstance(degree, int) and degree > 0):
+            raise ValueError("Argument `degree` expected to be integer larger than 0")
+        self.degree = degree
+
+        if gamma is not None and not isinstance(gamma, float):
+            raise ValueError("Argument `gamma` expected to be `None` or float larger than 0")
```
Comment (justusschock) [inline torchmetrics/image/kid.py:232]:
```suggestion
            raise TypeError("Argument `gamma` expected to be `None` or float larger than 0")
```

also reflect in docstrings :)

File: torchmetrics/image/kid.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+            extracted features.
+
+            Implementation inspired by https://github.com/toshas/torch-fidelity/blob/v0.3.0/torch_fidelity/metric_kid.py
+        """
+        real_features = torch.cat(self.real_features, dim=0)
+        fake_features = torch.cat(self.fake_features, dim=0)
+
+        n_samples_real = real_features.shape[0]
+        if n_samples_real < self.subset_size:
+            raise ValueError()
```
Comment (justusschock) [inline torchmetrics/image/kid.py:269]: Can you add a meaningful description here? :)

File: torchmetrics/image/kid.py [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+        n_samples_fake = fake_features.shape[0]
+        if n_samples_fake < self.subset_size:
+            raise ValueError()
```
Comment (justusschock) [inline torchmetrics/image/kid.py:272]: Meaningful description is missing here as well :)

**Review by justusschock — APPROVED** [review]
Some minor comments. Looks good overall

File: tests/image/test_kid.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    for _ in range(2):
+        img = torch.randint(0, 255, (10, 3, 299, 299), dtype=torch.uint8)
+        metric.update(img, real=True)
+        metric.update(img, real=False)
+
+    assert torch.allclose(torch.cat(metric.real_features, dim=0), torch.cat(metric.fake_features, dim=0))
+
+    mean, std = metric.compute()
+    assert mean != 0.0
+    assert std != 0.0
```
Comment (Borda) [inline tests/image/test_kid.py:114]: shall be at least set as positive or negative?

File: torchmetrics/image/kid.py [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+    kt_xx_sums = k_xx.sum(dim=-1) - diag_x
+    kt_yy_sums = k_yy.sum(dim=-1) - diag_y
+    k_xy_sums = k_xy.sum(dim=0)
+
+    kt_xx_sum = kt_xx_sums.sum()
+    kt_yy_sum = kt_yy_sums.sum()
+    k_xy_sum = k_xy_sums.sum()
+
+    value = (kt_xx_sum + kt_yy_sum) / (m * (m - 1))
+    value -= 2 * k_xy_sum / (m * m)
```
Comment (Borda) [inline torchmetrics/image/kid.py:43]:
```suggestion
    value -= 2 * k_xy_sum / (m ** 2)
```

File: torchmetrics/image/kid.py [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+
+        degree:
+            Degree of the polynomial kernel function
+
+        gamma:
+            Scale-length of polynomial kernel. If set to ``None`` will be automatically set to the feature size
+
+        coef:
+            Bias term in the polynomial kernel.
+
```
Comment (Borda) [inline torchmetrics/image/kid.py:121]:
```suggestion
        subset_size:
            Number of randomly picked samples in each subset
        degree:
            Degree of the polynomial kernel function
        gamma:
            Scale-length of polynomial kernel. If set to ``None`` will be automatically set to the feature size
        coef:
            Bias term in the polynomial kernel.
```

**Comment by Borda** [comment]
@SkafteNicki there seems to some argument errors with Conda tests

File: torchmetrics/image/kid.py (hunk 9 above)
Comment (justusschock) [reply torchmetrics/image/kid.py]: @Borda some time ago I saw some benchmarks that multiplication is usually faster for small powers then the actual power operation... Don't remember where though

File: torchmetrics/image/kid.py (hunk 7 above)
Comment (SkafteNicki) [reply torchmetrics/image/kid.py]: Done :]

File: torchmetrics/image/kid.py (hunk 6 above)
Comment (SkafteNicki) [reply torchmetrics/image/kid.py]: Done :]

File: tests/image/test_kid.py (hunk 8 above)
Comment (toshas) [reply tests/image/test_kid.py:120]: KID mean is allowed to be (slightly) negative for very close distributions

File: torchmetrics/image/kid.py [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+        rank_zero_warn(
+            'Metric `KID` will save all extracted features in buffer.'
+            ' For large datasets this may lead to large memory footprint.', UserWarning
+        )
+
+        if isinstance(feature, (str, int)):
+            if not _TORCH_FIDELITY_AVAILABLE:
+                raise RuntimeError(
+                    'KID metric requires that Torch-fidelity is installed.'
+                    'Either install as `pip install torchmetrics[image-quality]`'
```
Comment (Borda) [inline torchmetrics/image/kid.py:199]:
```suggestion
                    'Either install as `pip install torchmetrics[image]`'
```

**Review by Borda — APPROVED** [review]
